### PR TITLE
[tx] Add importance_sampling and ppo loss functions

### PR DESCRIPTION
**Pull request by Guido1Alessandro1Trevisan**
Added ppo and sampling loss functions. Tested these endpoints, and they work correctly. Note that different loss functions can be executed within the same forward_backward batch using jax.vmap with jax.lax.switch.

### CODE REVIEW

File: skyrl-tx/tx/tinker/engine.py [hunk 1]
```diff
@@ -60,6 +60,19 @@ def round_up_seq_len(seq_len: int) -> int:
...
+def _cross_entropy_loss(target_logprobs, loss_mask, sampling_logprobs, advantages):
+    return -target_logprobs * loss_mask
+
+def _importance_sampling_loss(target_logprobs, loss_mask, sampling_logprobs, advantages):
+    prob_ratio = jnp.exp(target_logprobs - sampling_logprobs)
+    return -(prob_ratio * advantages * loss_mask)
+
+def _ppo_loss(target_logprobs, loss_mask, sampling_logprobs, advantages):
+    prob_ratio = jnp.exp(target_logprobs - sampling_logprobs)
+    clipped_ratio = jnp.clip(prob_ratio, 0.8, 1.2)
```
[the remark this answers is not in the document]
Comment (Guido1Alessandro1Trevisan) [reply skyrl-tx/tx/tinker/engine.py]: Tinker states in the docs https://tinker-docs.thinkingmachines.ai/losses that the hardcoded epsilon is 0.2

File: skyrl-tx/tx/tinker/engine.py [hunk 2]
```diff
@@ -354,23 +393,46 @@ def process_forward_backward_batch(
...
+                    [1.0] * len(target_tokens),
+                    [0.0] * len(target_tokens),
+                    [0.0] * len(target_tokens),
+                )
+                if "weights" in loss_fn_inputs:
+                    weights = loss_fn_inputs["weights"]["data"]
+                if "logprobs" in loss_fn_inputs:
+                    sampling_lps = loss_fn_inputs["logprobs"]["data"]
+                if "advantages" in loss_fn_inputs:
+                    advs = loss_fn_inputs["advantages"]["data"]
```
[the remark this answers is not in the document]
Comment (Guido1Alessandro1Trevisan) [reply skyrl-tx/tx/tinker/engine.py]: I think this looks a bit ugly.  I intentionally went with the other design, but let me know what you think

File: skyrl-tx/tx/tinker/engine.py [hunk 3]
```diff
@@ -155,14 +168,34 @@ def loss_for_lora(
...
             logits = model(input_ids, attention_mask=attention_mask, adapter_indices=adapter_indices)[
                 "logits"
             ]  # [B, T, V]
-            per_token_losses = optax.softmax_cross_entropy_with_integer_labels(
-                logits=logits, labels=target_ids, where=loss_mask
-            )  # [B, T]
+            logprobs = jax.nn.log_softmax(logits, axis=-1)  # [B, T, V]
+            target_logprobs = jnp.take_along_axis(logprobs, target_ids[..., None], axis=-1).squeeze(-1)
+
+            loss_mask = jnp.where(loss_mask, 1.0, 0.0).astype(jnp.float32)
```
[the remark this answers is not in the document]
Comment (Guido1Alessandro1Trevisan) [reply skyrl-tx/tx/tinker/engine.py]: You got me

**Comment by pcmoritz** [comment]
I was trying to train a model and it pretty reliably produces NaN losses after a few steps on this branch, I'll do some more digging about what is wrong tomorrow :)
